### PR TITLE
create book entity, dao, service

### PR DESCRIPTION
**Pull request by Parashchuk**

### CODE REVIEW

File: src/main/java/bookstore/OnlineBookStoreApplication.java [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
+    private BookService bookService;
 
     public static void main(String[] args) {
         SpringApplication.run(OnlineBookStoreApplication.class, args);
     }
 
+    @Bean
+    public CommandLineRunner commandLineRunner() {
+        return args -> {
+            bookService.findAll();
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/OnlineBookStoreApplication.java:31]: you should add some data before findAll

File: src/main/java/bookstore/entity/Book.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import java.math.BigDecimal;
+import java.util.Objects;
+
+@Entity
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/Book.java:13]: Let’s  add Table annotation

File: src/main/java/bookstore/entity/Book.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Column(nullable = false, unique = true)
+    private String isbn;
+
+    @Column(nullable = false)
+    private BigDecimal price;
+
+    private String description;
+
+    private String coverImage;
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/Book.java]:
```suggestion
    private BigDecimal price;
    private String description;
    private String coverImage;
```

File: src/main/java/bookstore/entity/Book.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        return title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
+
+    public String getAuthor() {
+        return author;
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/entity/Book.java]: You can add Lombok and use annotations Getter, Setter, ToString etc.

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.List;
+import org.hibernate.SessionFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class BookRepositoryImpl implements BookRepository {
+    private final SessionFactory factory;
+
+    @Autowired
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]:
```suggestion
```
this annotation redundant from spring V4

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    public BookRepositoryImpl(SessionFactory factory) {
+        this.factory = factory;
+    }
+
+    @Override
+    public Book save(Book book) {
+        factory.inTransaction(e -> e.persist(book));
+        return book;
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]: use try-catch  construction (apply to all methods

File: src/main/java/bookstore/service/BookServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import bookstore.repository.BookRepositoryImpl;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class BookServiceImpl implements BookService {
+    private BookRepositoryImpl bookRepository;
+
+    @Autowired
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/BookServiceImpl.java]: the same

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  fix a few mistakes ;)

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    private final SessionFactory factory;
+
+    @Override
+    public Book save(Book book) {
+        try {
+            factory.inTransaction(e -> e.persist(book));
+        } catch (Exception e) {
+            throw new RuntimeException("Save operation was unsuccessful: ", e);
+        }
+
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]: remove empty line

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        return book;
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try {
+            return factory.fromSession(e -> e.createQuery("from Book", Book.class)
+                    .getResultList());
+        } catch (Exception e) {
+            throw new RuntimeException("Save operation was unsuccessful: ", e);
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]: It's bad practice to throw RuntimeException. Let's create custom DataProcessingException

Don't just copy-paste the message. 

"Can't find books" or "Can't get books from DB" will be better

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+@RequiredArgsConstructor
+public class BookRepositoryImpl implements BookRepository {
+    private final SessionFactory factory;
+
+    @Override
+    public Book save(Book book) {
+        try {
+            factory.inTransaction(e -> e.persist(book));
+        } catch (Exception e) {
+            throw new RuntimeException("Save operation was unsuccessful: ", e);
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]: It's bad practice to throw RuntimeException. Let's create custom DataProcessingException
The message could be "Can't save book to DB"

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        return book;
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java:27]: It's better to use try with resources here

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Well done! Let's improve your solution!

File: src/main/java/bookstore/repository/BookRepositoryImpl.java (hunk 11 above)
Comment (Elena-Bruyako) [reply src/main/java/bookstore/repository/BookRepositoryImpl.java:27]: Not fixed

**Review by Elena-Bruyako — APPROVED** [review]
Good job! Please fix one comment!

**Pull request merged**